v1/plugin/stream_collector_proxy.go

```diff
@@ -2,86 +2,130 @@ package plugin
+
+	// maxCollectionDuration sets the maximum duration between collections
+	// before metrics should be sent (defaults to 10s what means that after
+	// 10 seconds, the plugin should send data whatever it has instead of waiting longer)
+	maxCollectDuration time.Duration
 }
 
-func (c *StreamProxy) StreamMetrics(stream rpc.StreamCollector_StreamMetricsServer) error {
+func (p *StreamProxy) StreamMetrics(stream rpc.StreamCollector_StreamMetricsServer) error {
+	// set default max collect duration
+	p.setMaxCollectDuration(time.Duration(defaultMaxCollectDuration))
+
 	// Error channel where we will forward plugin errors to snap where it
 	// can report/handle them.
 	errChan := make(chan string)
 	// Metrics into the plugin from snap.
 	inChan := make(chan []Metric)
 	// Metrics out of the plugin into snap.
 	outChan := make(chan []Metric)
-	err := c.plugin.StreamMetrics(inChan, outChan, errChan)
+
+	err := p.plugin.StreamMetrics(inChan, outChan, errChan)
 	if err != nil {
 		return err
 	}
-	go metricSend(c.plugin, outChan, stream)
-	go errorSend(c.plugin, errChan, stream)
-	streamRecv(c.plugin, inChan, stream)
+
+	go p.metricSend(outChan, errChan, stream)
+	p.streamRecv(inChan, stream)
+
 	return nil
 }
 
-func errorSend(
-	plugin StreamCollector,
-	ch chan string,
-	s rpc.StreamCollector_StreamMetricsServer) {
-	for r := range ch {
-		reply := &rpc.CollectReply{
-			Error: &rpc.ErrReply{Error: r},
-		}
-		if err := s.Send(reply); err != nil {
-			fmt.Println(err.Error())
-		}
+func sendReply(metrics []*rpc.Metric, stream rpc.StreamCollector_StreamMetricsServer) {
+	if len(metrics) == 0 {
+		fmt.Println("No metrics available")
+		return
+	}
+
+	reply := &rpc.CollectReply{
+		Metrics_Reply: &rpc.MetricsReply{Metrics: metrics},
+	}
 
+	if err := stream.Send(reply); err != nil {
+		fmt.Println(err.Error())
 	}
 }
 
-func metricSend(
-	plugin StreamCollector,
-	ch chan []Metric,
-	s rpc.StreamCollector_StreamMetricsServer) {
-	for r := range ch {
-		mts := []*rpc.Metric{}
-		for _, mt := range r {
-			metric, err := toProtoMetric(mt)
-			if err != nil {
+func (p *StreamProxy) metricSend(metricsChan chan []Metric, errChan chan string, stream rpc.StreamCollector_StreamMetricsServer) {
+	metrics := []*rpc.Metric{}
+
+	for {
+		select {
+		case mts := <-metricsChan:
+			if len(mts) == 0 {
+				break
+			}
+
+			for _, mt := range mts {
+				metric, err := toProtoMetric(mt)
+				if err != nil {
+					fmt.Println(err.Error())
+					break
+				}
+				metrics = append(metrics, metric)
+
+				// send metrics if maxMetricsBuffer is reached
+				// notice, that only possible if maxMetricsBuffer is greater than 0
+				if p.maxMetricsBuffer == int64(len(metrics)) {
+					sendReply(metrics, stream)
+					metrics = []*rpc.Metric{}
+				}
+			}
+
+			// send all available metrics immediately for maxMetricsBuffer is 0 (defaults)
+			if p.maxMetricsBuffer == 0 {
+				sendReply(metrics, stream)
+				metrics = []*rpc.Metric{}
+			}
+
+		case err := <-errChan:
```

Comment:
As I talked with @andrzej-k, it will be better to have to receive data from errChan in a separate goroutine. So, I revert my changes according to this one.